File: 0079-word-search/0079-word-search.java
class Solution {
    int[][] directions = {{-1,0},{1,0},{0,1},{0,-1}};
    int m, n;
    public boolean exist(char[][] board, String word) {
        m = board.length;
        n = board[0].length;
        boolean[][] visited = new boolean[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (dfs(board, visited, i, j, 0, word)) return true;
            }
        }
        return false;
    }
    public boolean dfs(char[][] board, boolean[][] visited, int r, int c, int idx, String word) {
        // System.out.printf("checking %d, %d to see if board[%d][%d] equals %c \n", r, c, r, c, word.charAt(idx));
        // System.out.printf("idx = %d\n", idx);
        if (board[r][c]  !=  word.charAt(idx)) return false;
        if (idx == word.length() - 1) return true; //마지막 인덱스면 완성
        visited[r][c] = true;
        // System.out.printf("Matched!visited %c\n", board[r][c]);
        for (int[] d : directions) {
            int nr = r + d[0];
            int nc = c + d[1];
            if (nr < 0 || nr >= m || nc < 0 || nc >= n) continue;
            if (visited[nr][nc]) continue;
            
            //System.out.printf("Filter passed, checking adjacent directions for %d %d \n", r, c);
            if (dfs(board, visited, nr, nc, idx + 1, word)) return true;
            else continue;
        }
        visited[r][c] = false;
        return false;
    }
}

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        ROWS = len(board)
        COLS = len(board[0])
        directions = [[-1,0],[1,0],[0,1],[0,-1]]
        def dfs(r, c, board, word, idx):
            if r < 0 or r >= ROWS or c < 0 or c >= COLS or board[r][c] == '0': return False
            if board[r][c] != word[idx]: return False
            if idx == len(word) -1: return True #마지막 인덱스이고 동일하다면 종료
            #print(f"r={r},c={c}, idx={idx}")
            temp = board[r][c]
            board[r][c] = '0' #방문처리
            for d in directions:
                nr = r + d[0]
                nc = c + d[1]
                if dfs(nr,nc,board,word,idx+1): return True
            board[r][c] = temp
            
            return False
        
        for i in range(ROWS):
            for j in range(COLS):
                if dfs(i,j, board, word,0): return True
        return False
            

            